src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,126 @@ private int SkipBlockComment(int i)
+                case 'r': return "\r";
+                case 't': return "\t";
+                case 'u': return ScanUnicodeEscapeSequence();
+                case 'v': return "\v";
+                default: return c.ToString();
+            }
+        }
+
+        private string ScanUnicodeEscapeSequence()
+        {
+            int escSeqStartIndex = _currentIndex - 2;
+            int maxNumberOfHexDigits = 6;
+            bool isTerminated = false;
+
+            char c = GetChar();
+            if (c != '{')
+            {
+                UngetChar();
+
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                return s_unknownUnicodeChar;
+            }
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            int i;
+            for (i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Sequence has been terminated.
+                if (c == '}')
+                {
+                    if (i == 0)
+                    {
+                        // Sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                        return s_unknownUnicodeChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    ReportError(_currentIndex, () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (!isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex,
+                        i == maxNumberOfHexDigits && c.IsHexDigit()
+                            ? (Expression<Func<string>>)(() => ParserStrings.TooManyDigitsInUnicodeEscapeSequence)
+                            : () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
```

Comment:
@lzybkr:

All good points (and the problems I thought might arise from allowing surrogates were based on misconceptions).

The following is optional background information (it was news to me, so I thought I'd share it).

To my astonishment, **it is indeed C# / .NET itself that leaks the implementation details of UTF-16**:

```csharp
"\U0001F44D".Length   # !! 2 - leaking the internal representation as 2 surrogates
```

Since Unicode character `U+1F44D` cannot be represented as a _single_ UTF-16 code point / unit, it must _internally_ be represented as surrogate pair `U+D83D`, `U+DC4D`

So far, so good. However, that this internal representation is reflected in the _string length_ is unfortunate: **`.Length` counts the number of _code points_ rather than the number of _characters_** - which is akin to counting _bytes_ rather than characters in a UTF-8 string.

Note that **most of the aforementioned shells / scripting languages do _not_ exhibit this problematic behavior**, except, ironically, Python _2.x_ - the only one that handles explicit surrogates correctly - which acts like C#.

```sh
# Bash4, ksh, zsh
$ bash4 -c 'v=$(printf "\U0001F44D"); printf "$v\n"; printf "${#v}\n"'
👍
1

# Ruby
$ ruby -e 'v= "\u{1F44D}"; puts v, v.length'
👍
1

# Python 2.x
$ IOPYTHONENCODING=UTF-8 python -c 'v = u"\U0001F44D"; print(v); print(len(v))'
👍
2   # !! *2*, as in C#

# Python 3.x
$ python3 -c 'v = u"\U0001F44D"; print(v); print(len(v))'
👍
1

# Perl
$ perl -C -E  '$v="\N{U+1F44D}"; say $v; say length $v'
👍
1
```